Avoid brittle flags encoding for Java enums

som-snytt opened the pull request:
Fixes scala/bug#12800

In the current scheme, an enum should be `sealed abstract` and have only its elements as "children".

If an enum element has a body, then permittedSubclasses includes the element subclass.

The two tweaks are: don't add those permittedSubclasses as children; set `sealed abstract` unconditionally (on parsing an enum field, where the enum is already marked sealed).

The symptoms were that patmat would warn about `MyEnum$1` (where the anon subclass is a child) or `MyEnum forSome not in X,Y,Z` (if the enum is not marked abstract).

Second commit expunges the kludge of making enum always abstract just for exhaustivity. Instead, patmat analysis checks for an enum class (`isSealedEnum`). Class file parser assumes only present flags, except `ACC_ENUM` is translated to include `SEALED` for classes. Java parser checks enum members for a deferred member, and sets `ABSTRACT` accordingly.

som-snytt committed:
Set enum as abstract always

som-snytt committed:
Unwind abstract enum kludge

SethTisue commented:
> Second commit expunges the kludge of making enum always abstract just for exhaustivity

Is that kludge something we've had since forever, or something that was added recently?

I guess I'm asking partly out of curiosity and partly because I'm wondering what the risk level is of making a code improvement that isn't strictly necessary to fix bug (right?).

In other words, my "shouldn't this be two PRs" finger is twitching slightly, but my "Seth don't be pedantic" finger is twitching in the direction of the first finger, if the metaphor isn't getting too tortured.

SethTisue commented:
To overthink a little further, I guess whether I care depends on whether 2.13.12 ends up being a longer-incubating, optimistically-merge-things type release (as 2.13.11 was), or whether it ends up being an "oh shit we need to roll another one pronto" type release (as 2.13.10 was). And we don't know that yet.

som-snytt commented:
@SethTisue I expect @lrytz to weigh in on the second commit, hopefully with a ringing endorsement. Otherwise it can be moved to a separate PR as an internal improvement.

lrytz commented on src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala:
```diff
@@ -305,24 +305,17 @@ abstract class BTypesFromSymbols[G <: Global](val global: G) extends BTypes {
       ( if (classSym.isFinal)                                      ACC_FINAL     else 0) |
        // see the link above. javac does the same: ACC_SUPER for all classes, but not interfaces.)
       ( if (classSym.isInterface)                                  ACC_INTERFACE else ACC_SUPER) |
        // for Java enums, we cannot trust `hasAbstractFlag` (see comment in enumFlags))
```
```suggestion
```

som-snytt replied:
Thanks, I forgot to return to this spot. See the funky formatting of the next line.

lrytz commented on src/reflect/scala/reflect/internal/ClassfileConstants.scala:
```diff
@@ -353,7 +353,7 @@ object ClassfileConstants {
       case JAVA_ACC_ABSTRACT   => if (isClass) ABSTRACT else DEFERRED
       case JAVA_ACC_INTERFACE  => TRAIT | INTERFACE | ABSTRACT
-      case JAVA_ACC_ENUM       => JAVA_ENUM
+      case JAVA_ACC_ENUM       => if (isClass) JAVA_ENUM | SEALED else JAVA_ENUM
```
Here I don't follow, when is `isClass` true, when is it false?

som-snytt replied:
The enum member fields get the flag. (I believe I did not imagine that.)

lrytz reviewed (commented):
LGTM, both commits, just the `isSealedEnum` extension method seems a bit of an overkill.

som-snytt commented:
Updated with two suggestions.

In the beginning, I expected to be calling `isBlahBlah` all over the place, but I see it wound up as just a tweak to two expressions.

That is not in proportion to the degree of suffering required.

Also adjusted indentation on the flag translation, so that it fits in my default terminal.

I did not squash the two commits, but I did squash the two amendments. Dangerous to do before finishing first coffee! After deleting the package object, it tried to leak back into the second commit. ("How does git actually work again?")

lrytz reviewed (approved):
> Reluctantly renamed variable `abbie` to `hasAbstractFlag` (change not explicitly requested).

I like sentimental codebases
